### PR TITLE
PIM-9763: Make sure that 2 users can each create a private view with the same name

### PR DESCRIPTION
**Pull request by anaisbaunelemaire**
**Description (for Contributor and Core Developer)**

Before, you could only create public views.
These views were managed by a validation constraint which prevented the creation of a view with the same name as an already existing view (UniqueEntity constraint).

Recently, we can create private views.
Thus, a user can create a private view with the same name as the private or public view of another user.
On the other hand, a user cannot create a private view having the same name as another of his own private views.

Therefore, Pierre and I have created a new validator taking into account all these conditions.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Oro/Bundle/PimDataGridBundle/Validator/Constraints/UniqueDatagridViewEntityValidator.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            && $datagridViewFromDb->getId() !== $entity->getId()
+        ) {
+            $isLabelUnique = false;
+        }
+
+        if (!$isLabelUnique) {
+            $this->context->buildViolation($constraint->message)
+                ->atPath('label')
+                ->addViolation();
+        }
```
Comment (StevenVAIDIE) [inline src/Oro/Bundle/PimDataGridBundle/Validator/Constraints/UniqueDatagridViewEntityValidator.php:57]:
```suggestion
            $this->context->buildViolation($constraint->message)
                ->atPath('label')
                ->addViolation();
        }
```

File: src/Oro/Bundle/PimDataGridBundle/spec/Validator/Constraints/UniqueDatagridViewEntitySpec.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+namespace spec\Oro\Bundle\PimDataGridBundle\Validator\Constraints;
+
+use Oro\Bundle\PimDataGridBundle\Validator\Constraints\UniqueDatagridViewEntity;
+use PhpSpec\ObjectBehavior;
+use Symfony\Component\Validator\Constraint;
+
+class UniqueDatagridViewEntitySpec extends ObjectBehavior
```
Comment (StevenVAIDIE) [inline src/Oro/Bundle/PimDataGridBundle/spec/Validator/Constraints/UniqueDatagridViewEntitySpec.php:9]: Testing in phpspec constraint class is not necessary (for the next time)

**Review by StevenVAIDIE — APPROVED** [review]
I think you also need to change https://github.com/akeneo/pim-community-dev/blob/d589fcf1b9b15991e084ca32f18c27c4b6531885/src/Akeneo/UserManagement/Component/Updater/UserUpdater.php#L294 no ?

**Review by olivierpontier — APPROVED** [review]

File: src/Oro/Bundle/PimDataGridBundle/Validator/Constraints/UniqueDatagridViewEntity.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+use Symfony\Component\Validator\Constraint;
+
+/**
+ * @author Pierre Jolly <pierre.jolly@akeneo.com>
+ * @copyright 2021 Akeneo SAS (http://www.akeneo.com)
+ * @license http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class UniqueDatagridViewEntity extends Constraint
+{
+    public $message = 'The same label is already set on another view';
```
Comment (nmarniesse) [inline src/Oro/Bundle/PimDataGridBundle/Validator/Constraints/UniqueDatagridViewEntity.php:16]:
```suggestion
    public $message = 'pim_datagrid.column_configurator.label.unique_message';
```

And remove the message in *src/Oro/Bundle/PimDataGridBundle/Resources/config/validation/datagrid_view.yml*?

File: src/Oro/Bundle/PimDataGridBundle/Entity/DatagridView.php [hunk 4]
```diff
@@ -48,6 +48,13 @@ public function isPublic()
         return $this->type === self::TYPE_PUBLIC;
     }
 
+    public function setId(int $id): self
```
Comment (nmarniesse) [inline src/Oro/Bundle/PimDataGridBundle/Entity/DatagridView.php:51]: I would not add this method. The `id` is a critical property and must not be updated at any time.  
Please use a mock in the spec instead

File: src/Oro/Bundle/PimDataGridBundle/spec/Validator/Constraints/UniqueDatagridViewEntityValidatorSpec.php [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+        $datagridViewRepository->findPublicDatagridViewByLabel($datagridView->getLabel())
+            ->willReturn($datagridViewInDatabase);
+
+        $context->buildViolation('The same label is already set on another view')
+            ->willReturn($constraintViolationBuilder);
+        $constraintViolationBuilder->atPath('label')->willReturn($constraintViolationBuilder);
+        $constraintViolationBuilder->addViolation()->shouldBeCalled();
+
+        $this->validate($datagridView, $constraint)->shouldReturn(null);
```
Comment (nmarniesse) [inline src/Oro/Bundle/PimDataGridBundle/spec/Validator/Constraints/UniqueDatagridViewEntityValidatorSpec.php:54]: return is not important

```suggestion
        $this->validate($datagridView, $constraint);
```

**Review by nmarniesse — APPROVED** [review]
GTM after fix

File: src/Oro/Bundle/PimDataGridBundle/spec/Validator/Constraints/UniqueDatagridViewEntitySpec.php (hunk 2 above)
Comment (anaisbaunelemaire) [reply src/Oro/Bundle/PimDataGridBundle/spec/Validator/Constraints/UniqueDatagridViewEntitySpec.php:9]: Do you suggest that I delete the it_is_a_constraint() function?

Comment (nmarniesse) [reply src/Oro/Bundle/PimDataGridBundle/spec/Validator/Constraints/UniqueDatagridViewEntitySpec.php:9]: I think he suggests to remove the entire test class. I agree, there is no logic to test inside

**Comment by nmarniesse** [comment]
> I think you also need to change
> 
> https://github.com/akeneo/pim-community-dev/blob/d589fcf1b9b15991e084ca32f18c27c4b6531885/src/Akeneo/UserManagement/Component/Updater/UserUpdater.php#L294
> no ?

What do we change on this part plz?

**Comment by fitn** [comment]
It was green, I just resolve conflict on changelog

**Pull request merged**